README.md
# Miinaharava
Sovellus klassinen tietokonepeli Miinaharava. Käyttäjä voi pelata sovelluksella Miinaharavaa ja tallettaa huipputulokset tulostauluun.




## Dokumentaatio:  
- [vaatimusmäärittely](Projekti_miinaharava/Dokumentaatio/vaatimusmaarittely.md)  
- [työaikakirjanpito](Projekti_miinaharava/Dokumentaatio/tyoaikakirjanpito.md)  
- [arkkitehtuurikuvaus](Projekti_miinaharava/Dokumentaatio/arkkitehtuurikuvaus.md)   
- [changelog](changelog.md)  
- [käyttöohje](Projekti_miinaharava/Dokumentaatio/kayttoohje.md)

## Sovelluksen Käyttö

Käynnistettyäsi sovelluksen pääset menuun. Menussa voit määrittää kentän koon ja miinojen määrän klikkaamalla asetuksia ja kirjoittamalla määrän.  
Kentän koko voi olla välillä 5-20, miinoja voi olla vähintään 1, ja enintään puolet kentästä.  
Aloita peli painamalla 'Aloita', voit avata luukkuja painamalla niitä pelissä.  
Palaa menuun BACKSPACE-näppäimellä ja sule peli klikkaamalla 'Lopeta'

## Komentorivitoiminnot:

### Asennus
Lataa uusin julkaisu
- [Release Loppupalautus](https://github.com/EliasTHelsinginYliopisto/ot-harjoitustyo/releases/tag/v1.0.0)

1. 
```bash
poetry install
```
2.  
```bash
poetry run invoke build
```
3.  
```bash
poetry run invoke start
```

### Käynnistys

    poetry run invoke start
saatat joutua siirtymään 'Projekti_miinaharava' hakemistoon

### Testaus

    poetry run invoke test

### Testikattavuusreportin luonti

    poetry run invoke coverage-report
Testikattavuus luodaan htmlcov hakemistoon

### Pylint

    poetry run invoke lint


changelog.md
# Changelog

## Viikko 4

- Sovellus avaa menun
- Lisätty MainMenu-luokka joka määrittelee sovelluksen menun
- Lisätty GameTheme-luokka joka määrittää sovelluksen ulkonäön
- Lisätty Miinaharava-luokka joka kutsuu muita luokkia
- Testattu että menu käynnistyy

## Viikko 5

- Lisätty GameLogic-luokka joka suorittaa peliä määritetyilä asetuksilla
- Peli on nyt pelattavissa. lippuja ei voi vielä käyttää eikä peli pääty
- Testattu että peli käyttää annettuja asetuksia

## Viikko 6

- Miinoja voi nyt liputtaa
- Peli nyt asettaa oikean määrän miinoja
- Peli päättyy jos kaikki luukut on avattu tai käyttäjä avaa miinan
- Lisätty docstring

## Loppupalautus

- Korjattu bugi joka esti pelin aloittamisen oikeasta yläkulmasta

Projekti_miinaharava/src/tests/miinaharava_test.py
import unittest
import pygame
from miinaharava import Miinaharava


class TestMiinaharava(unittest.TestCase):
    def setUp(self):
        self._game = Miinaharava()

    def test_opens_menu_when_no_events(self):
        self._game.events()

        self.assertTrue(self._game.open_menu)

    """
    def test_opens_game_with_no_settings(self):
        self._game.events()
        pygame.mouse.set_pos([720/3, 150])
        mouse_pos = [720/3, 150]
        StubEvent(pygame.MOUSEBUTTONDOWN, mouse_pos)
        self._game._menu.events()
        pygame.time.delay(500)

        self.assertEqual(self._game._game.gridsize, 10)
    """

Projekti_miinaharava/src/tests/gameinstance_test.py
import unittest
import pygame
from gameinstance import GameLogic


class TestMiinaharava(unittest.TestCase):
    def setUp(self):
        self._game = GameLogic()
        self._game.window = pygame.display.set_mode(
            (1000, 1000)
        )

    def test_starts_without_settings(self):
        self._game.apply_settings(['', ''])
        self.assertEqual(self._game.gridsize, 10)
        self.assertEqual(self._game.mineamount, 20)
    
    def test_starts_with_invalid_settings(self):
        self._game.apply_settings([999, 999])
        self.assertEqual(self._game.gridsize, 20)
        self.assertEqual(self._game.mineamount, 200)


    def test_creates_grid_correctly(self):
        self._game.gridsize = 5
        grid1 = self._game.create_grid()
        grid2 = [[[0, 'closed'], [0, 'closed'], [0, 'closed'], [0, 'closed'], [0, 'closed']],
                 [[0, 'closed'], [0, 'closed'], [0, 'closed'],
                     [0, 'closed'], [0, 'closed']],
                 [[0, 'closed'], [0, 'closed'], [0, 'closed'],
                     [0, 'closed'], [0, 'closed']],
                 [[0, 'closed'], [0, 'closed'], [0, 'closed'],
                  [0, 'closed'], [0, 'closed']],
                 [[0, 'closed'], [0, 'closed'], [0, 'closed'], [0, 'closed'], [0, 'closed']]]
        self.assertEqual(grid1, grid2)

    def test_sets_correct_amount_of_mines(self):
        self._game.gridsize = 10
        self._game.mineamount = 91
        self._game.minefield = self._game.create_grid()
        start = [0, 0]
        self._game.set_mines(start)

        in_grid = 0
        for i in self._game.minefield:
            for j in i:
                if j[0] == 'M':
                    in_grid += 1
        self.assertEqual(self._game.mineamount, in_grid)
    
    def test_sets_mines_if_not_in_progress(self):
        self._game.gridsize = 5
        self._game.mineamount = 10
        self._game.minefield = self._game.create_grid()
        self._game.in_progress = False
        self._game.reveal_square(0, 0)



    def test_opening_square_sets_square_open(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = self._game.create_grid()
        self._game.window = pygame.display.set_mode(
            (1000, 1000)
        )
        self._game.reveal_square(0, 0)

        self.assertEqual(self._game.user_interface[2], 10)
    
    def test_cant_open_open_square(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = [[['0', 'open']]]
        self._game.window = pygame.display.set_mode(
            (1000, 1000)
        )
        self._game.reveal_square(0, 0)

        self.assertEqual(self._game.user_interface[2], 0)
   
   
    def test_opening_mine_ends_game(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = [[['M', 'closed']]]
        self._game.reveal_square(0, 0)

        self.assertEqual(self._game.game_state, 'over')
    

    def test_end_gives_flagging_bonus(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = [[['M', 'flagged']]]
        self._game.end_game()
        
        self.assertEqual(self._game.user_interface[2], 100)
    
    def test_can_set_flags(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = [[['M', 'closed']]]
        self._game.set_flag(0, 0)

        self.assertEqual(self._game.minefield, [[['M', 'flagged']]])
    
    def test_can_remove_flags(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = [[['M', 'flagged']]]
        self._game.set_flag(0, 0)

        self.assertEqual(self._game.minefield, [[['M', 'closed']]])

    def test_does_not_flag_outside_grid(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = [[['0', 'closed']]]
        self._game.set_flag(0, 1)

        self.assertEqual(self._game.minefield, [[['0', 'closed']]])

    def test_does_not_flag_open(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.minefield = [[['0', 'open']]]
        self._game.set_flag(0, 0)

        self.assertEqual(self._game.minefield, [[['0', 'open']]])

    def test_game_gives_win_bonus(self):
        self._game.gridsize = 1
        self._game.in_progress = True
        self._game.mineamount = 0
        self._game.left = 1
        self._game.minefield = [[['0', 'closed']]]
        self._game.reveal_square(0, 0)

        self.assertEqual(self._game.user_interface[2], 1010)


laskarit/viikko2/unicafe/src/tests/kassapaate_test.py
import unittest


import unittest
from kassapaate import Kassapaate
from maksukortti import Maksukortti

class TestKassapaate(unittest.TestCase):
    def setUp(self):
        self.kassapaate = Kassapaate()
        self.maksukortti = Maksukortti(1000)
    
    def test_kassapaate_luotu_oikein(self):
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)
        self.assertEqual(self.kassapaate.edulliset, 0)
        self.assertEqual(self.kassapaate.maukkaat, 0)
    
    def test_kateisosto_toimii_edullisella(self):
        self.assertEqual(self.kassapaate.syo_edullisesti_kateisella(300), 60)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100240)
        self.assertEqual(self.kassapaate.edulliset, 1)
        self.assertEqual(self.kassapaate.syo_edullisesti_kateisella(200), 200)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100240)
        self.assertEqual(self.kassapaate.edulliset, 1)

    def test_kateisosto_toimii_maukkaalla(self):
        self.assertEqual(self.kassapaate.syo_maukkaasti_kateisella(500), 100)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100400)
        self.assertEqual(self.kassapaate.maukkaat, 1)
        self.assertEqual(self.kassapaate.syo_maukkaasti_kateisella(300), 300)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100400)
        self.assertEqual(self.kassapaate.maukkaat, 1)

    def test_korttiosto_toimii_edullisella(self):
        self.assertEqual(self.kassapaate.syo_edullisesti_kortilla(self.maksukortti),True)
        self.assertEqual(self.maksukortti.saldo, 760)
        self.assertEqual(self.kassapaate.edulliset, 1)
        self.kassapaate.syo_edullisesti_kortilla(self.maksukortti)
        self.kassapaate.syo_edullisesti_kortilla(self.maksukortti)
        self.kassapaate.syo_edullisesti_kortilla(self.maksukortti)
        self.assertEqual(self.kassapaate.syo_edullisesti_kortilla(self.maksukortti),False)
        self.assertEqual(self.maksukortti.saldo, 40)
        self.assertEqual(self.kassapaate.edulliset, 4)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)

    def test_korttiosto_toimii_maukkaalla(self):
        self.assertEqual(self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti),True)
        self.assertEqual(self.maksukortti.saldo, 600)
        self.assertEqual(self.kassapaate.maukkaat, 1)
        self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti)
        self.assertEqual(self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti),False)
        self.assertEqual(self.maksukortti.saldo, 200)
        self.assertEqual(self.kassapaate.maukkaat, 2)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)

    def test_kortin_lataus(self):
        self.kassapaate.lataa_rahaa_kortille(self.maksukortti, 1000)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 101000)
        self.assertEqual(self.maksukortti.saldo, 2000)
        self.assertEqual(self.kassapaate.lataa_rahaa_kortille(self.maksukortti, -10), None)
        



    

Projekti_miinaharava/src/themes.py
import pygame


class DefineColours():
    """Luokka, joka määrittää värit teema luokkaa varten"""

    def __init__(self):
        self.black = (0, 0, 0)
        self.dark_grey = (96, 96, 96)
        self.light_grey = (160, 160, 160)
        self.white = (255, 255, 255)


class GameTheme():
    """Luokka joka määrittää värit ja fontit peliä varten

    Attributes:
        define: teeman värit"""

    def __init__(self):
        self._define = DefineColours()

        self.background_colour = self._define.black

        self.text_colour = self._define.dark_grey

        self.button_colour = self._define.light_grey

        self.closed_colour = self._define.white

        self.text_font = pygame.font.SysFont('Arial', 40)

        self.mines_font = pygame.font.SysFont('Arial', 20)

        self.title_font = pygame.font.SysFont('Arial', 100)


Projekti_miinaharava/src/mainmenu.py
import sys
import pygame
from themes import GameTheme


pygame.init()


class MainMenu:
    """Luokka joka suorittaa pelin päävalikkoa
    Attributes:
        clock:
            pygamen kello
        theme:
            Sovelluksen värit ja fontit
        running:
            boolean muuttuja siitä onko menu käynnissä
        buttons:
            Pygamen rect luokat menun napeista. 0 = Aloita,
            1 = lopeta, 2 = tulostaulukko
        input_rects:
            pygamen rect luokat pelin asetuksia varten
            0 = koko, 1 = miinat
        self.inputs:
            asetuksiin syötetty tieto.
            0 = koko, 1 = miinat.
        settingsactive:
            Pitää muistissa mitkä tekstikentät ovat aktiivisia
        window:
            pygamen näyttö
    """

    def __init__(self):
        """Luokan konstruktori"""

        self.clock = pygame.time.Clock()

        self._theme = GameTheme()

        self.running = False

        self.buttons = [None, None, None]
        self.input_rects = [None, None]
        self.inputs = [str(''), str('')]

        self.settingsactive = [False, False]

        self.window = None

    def events(self):
        """Lukee käyttäjän syötteet

        Returns:
            'quit_game': jos painetaan 'lopeta'-nappia
            'start_game': Jos painetaan 'aloita'-nappia
        """
        for event in pygame.event.get():

            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()

            if event.type == pygame.MOUSEBUTTONDOWN:
                mouse_pos = pygame.mouse.get_pos()
                if self.buttons[1].rect.collidepoint(mouse_pos):
                    pygame.display.quit()
                elif self.buttons[0].rect.collidepoint(mouse_pos):
                    return "start_game"
                elif self.input_rects[0].rect.collidepoint(mouse_pos):
                    self.settingsactive = [True, False]
                elif self.input_rects[1].rect.collidepoint(mouse_pos):
                    self.settingsactive = [False, True]
                else:
                    self.settingsactive = [False, False]

            if event.type == pygame.KEYDOWN:
                unicode = event.unicode
                key = event.key
                self.inputs_edit(unicode, key)
        return 'continue'

    def inputs_edit(self, unicode, key):
        """Muokkaa asetuksien tietoja annetun syötteen mukaan"""

        num09 = "1234567890"
        if unicode in num09:
            if self.settingsactive[0] and len(self.inputs[0]) < 4:
                self.inputs[0] += unicode
            elif self.settingsactive[1] and len(self.inputs[1]) < 4:
                self.inputs[1] += unicode
        if key == pygame.K_BACKSPACE:
            if self.settingsactive[0]:
                self.inputs[0] = self.inputs[0][:-1]
            elif self.settingsactive[1]:
                self.inputs[1] = self.inputs[1][:-1]

    def run(self):
        """suorittaa päävalikkoa

        Args:
        window_size_x: näytön leveys
        window_size_y: näytön korkeus"""

        window_size_x = 720
        window_size_y = 480
        self.window = pygame.display.set_mode(
            (window_size_x, window_size_y))

        self.running = True

        while self.running:

            self.window.fill(self._theme.background_colour)
            self.draw_menu_elements()
            self.input_boxes()

            status = self.events()
            if status == "start_game":
                return self.inputs
            if status == 'quit_game':
                return status

            self.clock.tick(60)
            pygame.display.update()

    def draw_menu_elements(self):
        """piirtää menutekstin ja napit

        Args:
        buttonsize: määrittää nappejen koon
        titlemarginal: marginaali näytön yläreunasta otsikkoon
        buttonmarginal: marginaali yläreunasta nappeihin
        window_size_x: näytön leveys
        """

        buttonsize = 1.3
        titlemarginal = 20
        buttonmarginal = 140
        window_size_x = 720
        title_text = self._theme.title_font.render(
            'MIINAHARAVA', True, self._theme.text_colour)
        self.window.blit(
            title_text, (titlemarginal, titlemarginal))

        self.buttons[0] = Button(window_size_x/3-50, buttonmarginal,
                                 buttonsize, self.window, self._theme.button_colour)
        start_text = self._theme.text_font.render(
            'Aloita', True, self._theme.text_colour)
        self.window.blit(
            start_text, (window_size_x/3-40, buttonmarginal+10))

        self.buttons[1] = Button(2*window_size_x/3-50, buttonmarginal,
                                 buttonsize, self.window, self._theme.button_colour)
        quit_text = self._theme.text_font.render(
            'Poistu', True, self._theme.text_colour)
        self.window.blit(
            quit_text, (2*window_size_x/3-40, buttonmarginal+10))

    def input_boxes(self):
        """sijoittaa tekstikentät näytölle

        Args:
        buttonsize: määrittää tekstikenttien koon
        window_size_x: näytön leveys
        settingsmarginal: marginaali näytön yläreunasta asetuksiin"""

        buttonsize = 1
        window_size_x = 720
        settingsmarginal = 240

        sizeset_text = self._theme.text_font.render(
            'Koko:', True, self._theme.text_colour)
        self.window.blit(
            sizeset_text, (window_size_x/3-100, settingsmarginal+10))
        self.input_rects[0] = Button(window_size_x/3+10, settingsmarginal+10,
                                     buttonsize, self.window, self._theme.button_colour)
        sizes_input = self._theme.text_font.render(
            self.inputs[0], True, self._theme.text_colour)
        self.window.blit(
            sizes_input, (window_size_x/3+10, settingsmarginal+20))

        mines_text = self._theme.text_font.render(
            'Miinat:', True, self._theme.text_colour)
        self.window.blit(
            mines_text, (2*window_size_x/3-100, settingsmarginal+10))
        self.input_rects[1] = Button(2*window_size_x/3+30, settingsmarginal+10,
                                     buttonsize, self.window, self._theme.button_colour)

        mines_input = self._theme.text_font.render(
            self.inputs[1], True, self._theme.text_colour)
        self.window.blit(
            mines_input, (2*window_size_x/3+40, settingsmarginal+20))


class Button():
    """Luokka, jonka avulla lisätään nappeja menuun
    """
    def __init__(self, x_coord, y_coord, size, surface, colour):
        self.rect = pygame.draw.rect(
            surface, colour, pygame.Rect((x_coord, y_coord), (100*size, 50*size)))


Projekti_miinaharava/src/miinaharava.py
import sys
import pygame
from mainmenu import MainMenu
from gameinstance import GameLogic


class Miinaharava:
    """pelin pääluokka, joka kutsuu alaluokkia

    Attributes:
        menu: pelin päävalikko
        game: pelin logiikka
        open_menu: onko peli käynnissä
        exit_game: suljetaanko sovellus
        start_game: käynnistetäänkö peli annetuilla asetuksilla
        running: onko sovellus käynnissä
    """


    def __init__(self):
        """luokan konstruktori"""

        self._menu = MainMenu()
        self._game = GameLogic()
        self.open_menu = False
        self.exit_game = False
        self.start_game = False
        self.running = False

    def events(self):
        """Sulkee pelin"""
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                self.exit_game = True
            else:
                self.open_menu = True

    def run(self):
        """pääluokan suoritus

        Args:
            status:
            luokkien palauttamat tiedot. menu palauttaa pelin asetukset,
            pelin logiikka palauttaa palataanko menuun
        """
        pygame.display.set_caption("Miinaharava")
        self.running = True

        while self.running:
            self.events()

            if self.exit_game:
                sys.exit()
            elif self.open_menu:
                status = self._menu.run()
                if isinstance(status, list):
                    self.open_menu = False
                    self.start_game = True
                elif status == 'quit_game':
                    pygame.quit()

            elif self.start_game:
                status = self._game.run(status)
                if status == 'return_to_menu':
                    self.start_game = False

            pygame.display.update()


Projekti_miinaharava/src/gameinstance.py
from random import randint
import sys
import pygame
from themes import GameTheme



class GameLogic:
    """Luokka, joka suorittaa peliä annetuilla asetuksilla.
        Attributes:
            mineamount:
                Miinojen määrä kentällä
            gridsize:
                Miinakentän leveys
            window:
                pygamen näyttö
            user_interface:
                Määrittää pelin UI:ssä näytettävän tiedon:
                0 = miinojen määrä
                1 = lippujen määrä
                2 = pisteiden määrä
            clock:
                pygamen kello
            theme:
                Sovelluksen värit ja fontit
            minefield:
                lista johon talletetaan miinakentän tiedot.
                muotoa [[[0, 1]]] missä 0 = luukun sisältämä tieto
                ja 1 = luukun tila. 'closed', 'open' tai 'flagged'
            left:
                kentällä olevien aukaisemattomien luukkujen määrä

            game_state:
                muuttuja joka pitää muistissa missä vaiheessa peli on
                not started = peli ei ole alkanut
                started = peli on käynnissä
                over = peli päättynyt
            """

    def __init__(self):
        """Luokan kontsruktori."""
        self.mineamount = 0
        self.gridsize = 0

        self.window = None

        self.user_interface = [0, 0, 0]

        self.clock = pygame.time.Clock()

        self._theme = GameTheme()

        self.minefield = None

        self.left = 0

        self.game_state = 'none'

    def run(self, settings):
        """Suorittaa peliä annetuilla asetuksilla

            Args:
                settings: asetukset, joilla peiä suoritetaan

            Returns:
                "return_to_menu" jos palataan takaisin päävalikkoon
        """

        self.apply_settings(settings)

        self.user_interface = [self.mineamount, 0, 0]

        self.minefield = self.create_grid()

        self.left = self.gridsize ** 2

        self.window.fill(self._theme.background_colour)
        self.display_grid()

        self.game_state = 'not started'

        while True:
            self.update_ui()
            status = self.events()
            if status == 'return_to_menu':
                return 'return_to_menu'

            self.clock.tick(60)
            pygame.display.update()

    def display_grid(self):
        """Piirtää annetun miinakentän näytölle"""
        i = 0
        j = 0
        while i < self.gridsize:
            while j < self.gridsize:
                pygame.draw.rect(
                    self.window, self._theme.closed_colour, pygame.Rect((i*30, 60+j*30), (28, 28)))
                j += 1
            j = 0
            i += 1

    def events(self):
        """Lukee annetut syötteet

        returns:
            "return_to_menu" jos painetaan askelpalautinta
        """
        for event in pygame.event.get():

            if event.type == pygame.QUIT:
                self.game_state = 'over'
                pygame.quit()
                sys.exit()

            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_BACKSPACE:
                    self.game_state = 'over'
                    return 'return_to_menu'

            if event.type == pygame.MOUSEBUTTONDOWN and self.game_state != 'over':
                mouse_pos = pygame.mouse.get_pos()
                line = mouse_pos[0] // 30
                row = (mouse_pos[1]-60) // 30
                if event.button == 1:
                    self.reveal_square(line, row)
                elif event.button == 3:
                    self.set_flag(line, row)
        return 'continue'

    def set_flag(self, line, row):
        """Asettaa, tai poistaa lipun miinakenttään

        Args:
            line: miinakentän linja
            row: miinakentän rivi
        """

        if line < 0 or line >= self.gridsize or row < 0 or row >= self.gridsize:
            return
        if self.minefield[row][line][1] == 'open':
            return

        if self.minefield[row][line][1] == 'flagged':
            self.minefield[row][line][1] = 'closed'
            self.user_interface[1] -= 1
            pygame.draw.rect(
                self.window, self._theme.closed_colour, pygame.Rect((line*30, 60+row*30), (28, 28)))

        else:
            self.minefield[row][line][1] = 'flagged'

            self.user_interface[1] += 1

            text = self._theme.mines_font.render(
                'F', True, self._theme.text_colour
            )
            self.window.blit(text, (5+line*30, 65+row*30))

    def reveal_square(self, line, row):
        """Avaa luukun valituista koordinaateista. jos yhtään luukkua
        ei ole vielä avattu, luodaan miinakenttä.
        """

        if line < 0 or line >= self.gridsize or row < 0 or row >= self.gridsize:
            return
        if self.minefield[row][line][1] != 'closed':
            return
        if self.game_state == 'not started':
            self.game_state = 'started'
            self.set_mines([line, row])

        pygame.draw.rect(
            self.window, self._theme.button_colour, pygame.Rect((line*30, 60+row*30), (28, 28)))

        self.minefield[row][line][1] = 'open'
        self.left -= 1

        self.user_interface[2] += 10

        self.check_contents(line, row)

    def check_contents(self, line, row):
        """tarkistetaan luukun sisältö. luukku jonka arvo on 0 avaa
        vierekkäiset luukut. jos luukkuja on enää sama määrä kuin miinoja,
        lopetetaan peli ja annetaan bonus.
        jos osutaan miinaan, peli päättyy.
        """

        content = self.minefield[line][row][0]

        if content == 0:

            surrounding = [(row-1, line-1), (row-1, line), (row-1, line+1),
                           (row, line-1), (row, line+1),
                           (row+1, line-1), (row+1, line), (row+1, line+1)]
            for square in surrounding:
                self.reveal_square(square[1], square[0])
            return

        if content == 'M':
            self.user_interface[2] -= 10
            self.end_game()
            return

        text = self._theme.mines_font.render(
            str(content), True, self._theme.text_colour
        )
        self.window.blit(text, (5+line*30, 65+row*30))

        if self.left == self.mineamount:
            self.user_interface[2] += 1000
            self.end_game()

    def end_game(self):
        """Lopettaa pelin. jokaisesta liputetusta miinasta tulee 100 pisteen bonus.
        jos miinoja on jäljellä, niiden sijainnit paljastetaan.
        """

        self.game_state = 'over'

        for i in range(self.gridsize):
            for j in range(self.gridsize):
                if self.minefield[i][j][0] == 'M':
                    self.reveal_mine(i, j)

    def reveal_mine(self, i, j):
        """Piirtää miinan näytölle. jos miina on liputettu
        sitä ei piirretä, ja annetaan bonus
        """
        if self.minefield[j][i][1] == 'flagged':
            self.user_interface[2] += 100
        else:
            text = self._theme.mines_font.render(
                str('M'), True, self._theme.text_colour
            )
            self.window.blit(text, (5+i*30, 65+j*30))



    def set_mines(self, start):
        """Asettaa miinat satunnaisesti siten että aloitusruudun
        vieressä ei ole miinoja
        """

        width = self.gridsize
        start = start[0]*width + start[1]
        spots = width ** 2
        places = list(range(spots))
        start_protection = [start-width-1, start-width, start-width+1, start-1,
                            start, start+1, start+width-1, start+width, start+width+1]
        for spot in start_protection:
            if 0 <= spot <= spots-1:
                places.remove(spot)
        spots = spots-10

        i = 0
        while i < self.mineamount:
            placement = randint(0, spots)
            spot = places[placement]
            del places[placement]
            p_y = spot // width
            p_x = spot % width
            self.minefield[p_y][p_x][0] = 'M'
            self.increase_heat(p_y, p_x)

            spots = spots-1
            i += 1

    def increase_heat(self, p_y, p_x):
        """Kasvattaa miinan viereisten luukkujen numeroa"""

        width = self.gridsize

        increase_heat = [
            [p_x-1, p_y-1], [p_x-1, p_y], [p_x-1, p_y+1],
            [p_x, p_y-1], [p_x, p_y+1],
            [p_x+1, p_y-1], [p_x+1, p_y], [p_x+1, p_y+1]
            ]
        for spot in increase_heat:
            if spot[0] >= 0 and spot[0] < width and spot[1] >= 0 and spot[1] < width:
                if self.minefield[spot[1]][spot[0]][0] != "M":
                    self.minefield[spot[1]][spot[0]][0] += 1


    def create_grid(self):
        """Luo taulukon jonka koko on annettu leveys²

        Returns:
            grid1: Luotu taulukko
        """
        grid1 = []
        for _ in range(self.gridsize):
            grid2 = []
            for _ in range(self.gridsize):
                grid2.append([0, 'closed'])
            grid1.append(grid2)
        return grid1

    def update_ui(self):
        """Päivittää näytolle self.user_interface muuttujaan
        tehdyt muutokset"""

        if self.game_state != 'over':
            ui_text = self._theme.text_font.render(
                f"Miinoja: {self.user_interface[0]} | "
                f"Liput {self.user_interface[1]}| Pisteet {self.user_interface[2]} |",
                True, self._theme.text_colour
            )
        else:
            ui_text = self._theme.text_font.render(
                f"""Peli Päättynyt | Pisteet {self.user_interface[2]} |""",
                True, self._theme.text_colour
            )
        pygame.draw.rect(
            self.window, self._theme.background_colour, pygame.Rect(
                (0, 0), (700, 60))
        )
        self.window.blit(ui_text, (10, 10))

    def apply_settings(self, settings):
        """Tarkistaa onko peli aloitettu sopivilla asetuksilla
        ja korjaa ne sopivaan muotoon
        """

        if settings[0] == '':
            settings[0] = '10'
        if settings[1] == '' or settings[1] == 0:
            settings[1] = '20'

        self.gridsize = int(settings[0])
        self.mineamount = int(settings[1])

        self.gridsize = max(self.gridsize, 5)
        self.gridsize = min(self.gridsize, 20)

        if self.mineamount > (self.gridsize ** 2)*0.5:
            self.mineamount = int((self.gridsize ** 2)*0.5)

        window_x = 30*self.gridsize
        window_x = max(window_x, 700)
        self.window = pygame.display.set_mode(
            (window_x, 80+30*self.gridsize)
        )
